client/web/src/hooks/useScrollManager.ts

```diff
@@ -0,0 +1,164 @@
+import { useEffect } from 'react'
+
+import { debounce } from 'lodash'
+import { useLocation } from 'react-router'
+
+/**
+ * Object containing maps of saved scroll locations for each path name, which are keyed by the scrollable container they belong to.
+ * This will allow us to easily support scroll preservation in multiple containers, even when they are used in the same page.
+ */
+const SAVED_SCROLL_POSITIONS: { [k: string]: Map<string, number> } = {}
+
+/**
+ * Object containing maps of cached scroll attempts for each path name, which are keyed by the scrollable container they belong to.
+ * This can be used to cancel a scroll attempt if a new attempt comes in for the same path.
+ */
+const CACHED_SCROLL_ATTEMPTS: { [k: string]: Map<string, MutationObserverPromise> } = {}
+
+/** The length of the timeout in milliseconds until we stop trying to scroll. */
+const SCROLL_RETRY_TIMEOUT = 3000
+
+export function useScrollManager(containerKey: string, containerRef: React.RefObject<HTMLElement>): void {
+    const { pathname } = useLocation()
+
+    // Set up the maps for this containerKey if they haven't been created yet
+    useEffect(() => {
+        if (!SAVED_SCROLL_POSITIONS[containerKey]) {
+            SAVED_SCROLL_POSITIONS[containerKey] = new Map<string, number>()
+        }
+
+        if (!CACHED_SCROLL_ATTEMPTS[containerKey]) {
+            CACHED_SCROLL_ATTEMPTS[containerKey] = new Map<string, MutationObserverPromise>()
+        }
+    })
+
+    // Attach scroll listener, make sure to detach on unmount
+    useEffect(() => {
+        const container = containerRef.current
+
+        const saveScrollPosition = debounce(() => {
+            if (container) {
+                console.log(`Saving position '${container.scrollTop}' for pathname '${pathname}'`)
+                SAVED_SCROLL_POSITIONS[containerKey].set(pathname, container.scrollTop)
+            }
+        }, 200)
+
+        container?.addEventListener('scroll', saveScrollPosition)
+
+        return () => container?.removeEventListener('scroll', saveScrollPosition)
+    }, [pathname, containerKey, containerRef])
+
+    // Handle changes to pathname and try to scroll to the saved position at that path, if it exists
+    useEffect(() => {
+        // Cancel any existing cached scroll attempts for this pathname
+        if (CACHED_SCROLL_ATTEMPTS[containerKey].has(pathname)) {
+            CACHED_SCROLL_ATTEMPTS[containerKey].get(pathname)?.cancel()
+            CACHED_SCROLL_ATTEMPTS[containerKey].delete(pathname)
+        }
+
+        // Attempt a scroll if we have a saved position for the pathname; if the scroll doesn't work, set up an observer to
+        // retry up until a given timeout
+        if (SAVED_SCROLL_POSITIONS[containerKey].has(pathname)) {
```

Comment:
I think we only want to do this if the user is actively navigating history. It might be a bit confusing if users click a link and get auto-scrolled because they were on that same like 5 mins ago.

I think we should just be able to use `useHistory` and change this line to:

```suggestion
        if (history.action === 'POP' && SAVED_SCROLL_POSITIONS[containerKey].has(pathname)) {
```

Reply:
Pushed this up with my change to add tests, keeping open for discussion!